File: src/cnn.cpp
#include "cnn.h"

CNNImpl::CNNImpl() {
        conv1 = register_module("conv1", torch::nn::Conv2d(1, 16, 3));  // 255x255x1 -> 253x253x16
        bn1 = register_module("bn1", torch::nn::BatchNorm2d(16)); 
        relu1 = register_module("relu1", torch::nn::ReLU());  
        pool1 = register_module("pool1", torch::nn::MaxPool2d(2)); // 126x126x16

        conv2 = register_module("conv2", torch::nn::Conv2d(16, 32, 3)); // 124x124x32
        bn2 = register_module("bn2", torch::nn::BatchNorm2d(32));
        relu2 = register_module("relu2", torch::nn::ReLU());
        pool2 = register_module("pool2", torch::nn::MaxPool2d(2)); // 62x62x32

        flatten = register_module("flatten", torch::nn::Flatten()); // 32*62*62
        dropout = register_module("dropout", torch::nn::Dropout(0.5));
        fc1 = register_module("fc1", torch::nn::Linear(32 * 62 * 62, 14)); // 14 classes
        softmax = register_module("softmax", torch::nn::LogSoftmax(1));
    }

torch::Tensor CNNImpl::forward(torch::Tensor x){
        x = pool1->forward(relu1->forward(bn1->forward(conv1->forward(x))));
        x = pool2->forward(relu2->forward(bn2->forward(conv2->forward(x))));
        x = flatten->forward(x);
        x = dropout->forward(x);
        x = fc1->forward(x);
        x = softmax->forward(x);
        return x;
    }

File: scripts/cnn.py
from torch import nn

class CNN (nn.Module):
    """A Convolutional Neural Network (CNN) model for gesture recognition.

    The model consists of two convolutional blocks followed by a fully connected layer.
    It is designed to process grayscale images of gestures.
    """
    def __init__(self):
        """Initializes the CNN model layers.

        The architecture includes:
        - Two convolutional layers with Batch Normalization, ReLU activation, and Max Pooling.
        - A Flatten layer to convert 2D feature maps to a 1D vector.
        - A Dropout layer for regularization.
        - A final Fully Connected layer for classification into 14 classes.
        """
        super(CNN, self).__init__()
        self.conv1 = nn.Conv2d(1, 8, kernel_size=3) # 32x32x1 -> 30x30x8
        self.bn1 = nn.BatchNorm2d(8)
        self.relu1 = nn.ReLU()
        self.pool1 = nn.MaxPool2d(kernel_size=2) # 15x15x8

        self.conv2 = nn.Conv2d(8, 16, kernel_size=3) # 13x13x16
        self.bn2 = nn.BatchNorm2d(16)
        self.relu2 = nn.ReLU()
        self.pool2 = nn.MaxPool2d(kernel_size=2) # 6x6x16

        self.flatten = nn.Flatten()
        self.dropout = nn.Dropout(0.5)
        self.fc1 = nn.Linear(6*6*16, 14) # 14 classes

    def forward(self, x):
        """Defines the forward pass of the CNN model.

        Args:
            x (torch.Tensor): The input tensor, expected to be a grayscale image batch.

        Returns:
            torch.Tensor: The output tensor containing the logits for each class.
        """
        x = self.conv1(x)
        x = self.pool1(self.relu1(self.bn1(x)))
        x = self.pool2(self.relu2(self.bn2(self.conv2(x))))
        x = self.flatten(x)
        x = self.dropout(x)
        x = self.fc1(x)
        return x